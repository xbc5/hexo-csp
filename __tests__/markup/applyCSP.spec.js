"use strict";

const Config = require("../../lib/config");
const Policies = require("../../lib/Policies");
const { applyCSP } = require("../../lib/markup");
const template = require("../helpers/template");

function fixture(conf, data, str, path, err) {
  const c = new Config(conf);
  const policies = new Policies();
  // policies are combined when the app is initialised, so combine them
  // with savePolicy to emulate this.
  conf?.policies?.forEach((p) => policies.savePolicy(p));
  return applyCSP(c, policies, data, str, path, err);
}

// TODO:
// - test dev env;
//
describe("given valid args", () => {
  it("should return the markup with CSP applied", async () => {
    const c = {
      policies: [
        {
          prod: {
            pattern: "foo",
            directives: {
              "default-src": ["'self'", "https://foo.com"],
              "img-src": ["https://bar.com"],
            },
          },
        },
      ],
    };
    const data = { path: "foo/index.html" };

    const markup = fixture(c, data, template.noMeta, data.path);

    expect(markup).toMatchSnapshot();
  });

  it("should apply only the corresponding polices for that path (foo/index.html", async () => {
    const c = {
      policies: [
        {
          pattern: "^bar$",
          prod: {
            directives: {
              "default-src": ["https://config-default-default.com"],
            },
          },
        },
        {
          pattern: "^foo",
          prod: {
            directives: {
              "default-src": ["https://config-foo-default.com"],
            },
          },
        },
      ],
    };
    const data = {
      path: "foo/index.html",
      page: {
        csp: {
          prod: {
            mode: "merge",
            directives: {
              "default-src": ["https://frontmatter-foo-default.com"],
            },
          },
        },
      },
    };

    const markup = fixture(c, data, template.noMeta, data.path);

    expect(markup).toMatchSnapshot();
  });
});

describe("given frontmatter", () => {
  describe("mode=[default] (merge)", () => {
    it("should merge the frontmatter and return the markup", async () => {
      const c = {
        policies: [
          {
            pattern: "foo",
            prod: {
              directives: {
                "default-src": ["https://config-foo-default.com"],
                "img-src": ["https://config-foo-img.com"],
              },
            },
          },
        ],
      };
      const data = {
        path: "foo/index.html",
        page: {
          csp: {
            prod: {
              // FIXME: pattern?
              directives: {
                "default-src": ["https://frontmatter-foo-default.com"],
                "img-src": ["https://frontmatter-foo-img.com"],
              },
            },
          },
        },
      };

      const markup = fixture(c, data, template.noMeta, data.path);

      expect(markup).toMatchSnapshot();
    });
  });

  describe("mode=merge", () => {
    it("should merge the frontmatter and return the markup", async () => {
      const c = {
        policies: [
          {
            pattern: "foo",
            prod: {
              directives: {
                "default-src": ["https://config-foo-default.com"],
                "img-src": ["https://config-foo-img.com"],
              },
            },
          },
        ],
      };
      const data = {
        path: "foo/index.html",
        page: {
          csp: {
            // FIXME: pattern?
            prod: {
              mode: "merge",
              directives: {
                "default-src": ["https://frontmatter-foo-default.com"],
                "img-src": ["https://frontmatter-foo-img.com"],
              },
            },
          },
        },
      };

      const markup = fixture(c, data, template.noMeta, data.path);

      expect(markup).toMatchSnapshot();
    });
  });

  describe("mode=replace", () => {
    it("should replace the config with the frontmatter and return the markup", async () => {
      const c = {
        policies: [
          {
            // applyCSP encloses the path (minus index) with ^$:
            // foo/index.html => ^foo$, and uses that value to do a literal
            // comparison. If that comparison doesn't match, then no policies
            // will get replaced. So use ^foo$ so that the patterns are identical.
            pattern: "^foo$",
            prod: {
              directives: {
                "default-src": ["https://config-foo-default.com"],
                "img-src": ["https://config-foo-img.com"],
              },
            },
          },
        ],
      };
      const data = {
        path: "foo/index.html",
        page: {
          csp: {
            prod: {
              mode: "replace",
              directives: {
                "default-src": ["https://frontmatter-foo-default.com"],
                "img-src": ["https://frontmatter-foo-img.com"],
              },
            },
          },
        },
      };

      const markup = fixture(c, data, template.noMeta, data.path);

      expect(markup).toMatchSnapshot();
    });
  });
});

describe("given no directives to apply", () => {
  [
    undefined,
    null,
    {},
    { page: {} },
    { page: { policies: [] } },
    { page: { policies: [{ prod: {} }] } },
    { page: { policies: [{ prod: { directves: {} } }] } },
  ].forEach((c) => {
    const param = c === undefined ? undefined : JSON.stringify(c);
    describe(`i.e. the config object is ${param}`, () => {
      it("it should return the original markup", async () => {
        const data = { path: "foo/index.html" };
        const markup = fixture(c, data, template.noMeta, data.path);
        expect(markup).toBe(template.noMeta);
      });
    });
  });
});

describe("for inline tags", () => {
  describe("given inline scripts, styles and additional policies", () => {
    it("should merge content hashes, and apply policies", async () => {
      const c = {
        inline: {
          enabled: true,
          algo: "sha256",
        },
        policies: [
          {
            pattern: "foo",
            prod: {
              directives: {
                "style-src": ["https://policy-foo-style.com"],
                "script-src": ["https://policy-foo-script.com"],
                "default-src": ["https://policy-foo-default.com"],
              },
            },
          },
        ],
      };
      const data = { path: "foo/index.html" };

      const markup = fixture(
        c,
        data,
        template.mixedStyleScriptsForHashing,
        data.path
      );

      expect(markup).toMatchSnapshot();
    });
  });

  ["sha256", "sha384", "sha512"].forEach((algo) => {
    describe(`and given ${algo}`, () => {
      it(`it should inject a ${algo}+base64 source for each`, async () => {
        const c = {
          inline: {
            enabled: true,
            algo,
          },
        };
        const data = { path: "foo/index.html" };

        const markup = fixture(
          c,
          data,
          template.mixedStyleScriptsForHashing,
          data.path
        );

        expect(markup).toMatchSnapshot();
      });
    });
  });

  describe("when disabled", () => {
    it("should not generate hashes (but still apply other policies)", async () => {
      const c = {
        inline: {
          enabled: false,
          algo: "sha256",
        },
        policies: [
          {
            pattern: "foo",
            prod: {
              directives: {
                "style-src": ["https://policy-foo-style.com"],
                "script-src": ["https://policy-foo-script.com"],
                "default-src": ["https://policy-foo-default.com"],
              },
            },
          },
        ],
      };
      const data = { path: "foo/index.html" };

      const markup = fixture(
        c,
        data,
        template.mixedStyleScriptsForHashing,
        data.path
      );

      expect(markup).toMatchSnapshot();
    });
  });
});
