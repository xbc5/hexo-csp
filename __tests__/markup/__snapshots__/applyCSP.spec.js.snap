// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`for inline tags and given sha256 it should inject a sha256+base64 source for each 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="script-src 'sha256-tuIk8UMM5VK6Dn8Jn3ePSolo3YszPi+JeJrdzrOshsw=' 'sha256-qnFkTcuoAp6oO1RZJy6aN/9o3iansAMJSx0ZLqbk/JY='; style-src 'sha256-nMd1F/LMKzbEm8mWbcTNC6Ifar1sutAMJJ1uqhvJ/c8=' 'sha256-hEATX7sUZolX9HQYaQ+sSRbknzWRnW8y7ZPdFlbaYd0='"
    >
    <title>
      Fake Title
    </title>
    <script>
      const a=1;
    </script>
    <style>
      div {
        css: sucks;
      }
    </style>
  </head>
  <body>
    <script>
      const b=1;
    </script>
    <style>
      div {
        css: is-bad;
      }
    </style>
  </body>
</html>
`;

exports[`for inline tags and given sha384 it should inject a sha384+base64 source for each 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="script-src 'sha384-uZT+XtG7RuJtephpOtu6rJHWMo4jkDXqF7zpKllBBnBCUDGEC1EAfUxa9mE7cdcb' 'sha384-xVuC9/ZsdCoMmOnFcN5205oZDIqRbOd7bcqNv7xyKsr7vgg1gh3MAc8MehJbJGBI'; style-src 'sha384-nPNwinjybwk0/Dyw1k3bDabl4blBnQ27bv67RSyTqgTlx+E4NK3N3kWmLjNab09K' 'sha384-2NucM7fXDV2sNuEglrlIpvSY6a3M2wNOzuw7Qjg7+J3hN5Lw8JG3SSLsSRwqWL6+'"
    >
    <title>
      Fake Title
    </title>
    <script>
      const a=1;
    </script>
    <style>
      div {
        css: sucks;
      }
    </style>
  </head>
  <body>
    <script>
      const b=1;
    </script>
    <style>
      div {
        css: is-bad;
      }
    </style>
  </body>
</html>
`;

exports[`for inline tags and given sha512 it should inject a sha512+base64 source for each 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="script-src 'sha512-/558eYLHKn8YKWknLoF5RvQwEOb05/pU4pFSzvN/0FW5vPSlHYILUisxdm/C5Ss1yQwD80DMaYASRhhK7JfArw==' 'sha512-2DIFEephSKwxrINwes/H9aLafWq5GfOZtEZUVP9gJD5qSLiazykCqpXu0cIX+ZholB5xAfT4wnfAhiVsBuy+Lg=='; style-src 'sha512-lmCFcZx5gkRkUB18igbCihWI/zlKQs0pbn6BEmtWQz01pCYA/anuvUrPSbMXiQtqBLkg7/cusKnL9VUdKwJyhw==' 'sha512-U7SEuKqtMdgBKGjsCu7ShMTmOIPyk4BH9L/L3WHSlCmJwcpTB2BxLN5O72tsgtHuvmoglf8ySNmNuaG7iavYiA=='"
    >
    <title>
      Fake Title
    </title>
    <script>
      const a=1;
    </script>
    <style>
      div {
        css: sucks;
      }
    </style>
  </head>
  <body>
    <script>
      const b=1;
    </script>
    <style>
      div {
        css: is-bad;
      }
    </style>
  </body>
</html>
`;

exports[`for inline tags given inline scripts, styles and additional policies should merge content hashes, and apply policies 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="style-src https://policy-foo-style.com 'sha256-nMd1F/LMKzbEm8mWbcTNC6Ifar1sutAMJJ1uqhvJ/c8=' 'sha256-hEATX7sUZolX9HQYaQ+sSRbknzWRnW8y7ZPdFlbaYd0='; script-src https://policy-foo-script.com 'sha256-tuIk8UMM5VK6Dn8Jn3ePSolo3YszPi+JeJrdzrOshsw=' 'sha256-qnFkTcuoAp6oO1RZJy6aN/9o3iansAMJSx0ZLqbk/JY='; default-src https://policy-foo-default.com"
    >
    <title>
      Fake Title
    </title>
    <script>
      const a=1;
    </script>
    <style>
      div {
        css: sucks;
      }
    </style>
  </head>
  <body>
    <script>
      const b=1;
    </script>
    <style>
      div {
        css: is-bad;
      }
    </style>
  </body>
</html>
`;

exports[`for inline tags when disabled should not generate hashes (but still apply other policies) 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="style-src https://policy-foo-style.com; script-src https://policy-foo-script.com; default-src https://policy-foo-default.com"
    >
    <title>
      Fake Title
    </title>
    <script>
      const a=1;
    </script>
    <style>
      div {
        css: sucks;
      }
    </style>
  </head>
  <body>
    <script>
      const b=1;
    </script>
    <style>
      div {
        css: is-bad;
      }
    </style>
  </body>
</html>
`;

exports[`given frontmatter mode=[default] (merge) should merge the frontmatter and return the markup 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://config-foo-default.com https://frontmatter-foo-default.com; img-src https://config-foo-img.com https://frontmatter-foo-img.com"
    >
    <title>
      Fake Title
    </title>
  </head>
  <body>
  </body>
</html>
`;

exports[`given frontmatter mode=merge should merge the frontmatter and return the markup 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://config-foo-default.com https://frontmatter-foo-default.com; img-src https://config-foo-img.com https://frontmatter-foo-img.com"
    >
    <title>
      Fake Title
    </title>
  </head>
  <body>
  </body>
</html>
`;

exports[`given frontmatter mode=replace should replace the config with the frontmatter and return the markup 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://frontmatter-foo-default.com; img-src https://frontmatter-foo-img.com"
    >
    <title>
      Fake Title
    </title>
  </head>
  <body>
  </body>
</html>
`;

exports[`given valid args should apply only the corresponding polices for that path (foo/index.html 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src https://config-foo-default.com https://frontmatter-foo-default.com"
    >
    <title>
      Fake Title
    </title>
  </head>
  <body>
  </body>
</html>
`;

exports[`given valid args should return the markup with CSP applied 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src 'self' https://foo.com; img-src https://bar.com"
    >
    <title>
      Fake Title
    </title>
  </head>
  <body>
  </body>
</html>
`;
